src/Hosting/TestHost/src/HttpContextBuilder.cs

```diff
@@ -134,8 +149,40 @@ internal void ClientInitiatedAbort()
                 // We don't want to trigger the token for already completed responses.
                 _requestLifetimeFeature.Cancel();
             }
+
             // Writes will still succeed, the app will only get an error if they check the CT.
             _responseReaderStream.Abort(new IOException("The client aborted the request."));
+
+            // Cancel any pending request async activity when the client aborts a duplex
+            // streaming scenario by disposing the HttpResponseMessage.
+            AbortRequest();
+        }
+
+        private async Task CompleteRequestAsync()
+        {
+            if (!_requestPipe.Reader.TryRead(out var result) || !result.IsCompleted)
+            {
+                // If request is still in progress then abort it.
+                AbortRequest();
+            }
+            else
+            {
+                await _requestPipe.Writer.CompleteAsync();
```

Comment:
Don't do this here, complete the writer on the code doing the writing. Completing the reading should be fine to do here before the task ends.